horovod/_keras/__init__.py

```diff
@@ -167,32 +185,83 @@ def allreduce_grads():
+
+            def profile_start():
+                if self._profile_frequency == 0:
+                    return tf.no_op()
+                return tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), log_comm_start, tf.no_op)
+
+            def log_comm_start():
+                profile_base_info = trim_last_curly_brace(get_profile_info("communication", "B"))
+                self.start_timestamp = tf.timestamp() * 1e6
+                return tf.print(profile_base_info, ", \"ts\": ", self.start_timestamp, "}", sep="", output_stream="file:///profile/pedl_profiling.log")
+
+            def get_profile_info(name, phase, **kwargs):
+                info = {
+                    "name": name,
+                    "pid": os.getpid(),
+                    "tid": threading.current_thread().ident,
+                    "ph": phase,
+                }
+                info.update(kwargs)
+                return json.dumps(info)
+
+            def profile_end():
+                if self._profile_frequency == 0:
+                    return tf.no_op()
+                cond = tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), log_comm_end, tf.no_op)
+                with tf.control_dependencies([cond]):
+                    return tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), clear_profile_counter, increment_profile_counter)
+
+            def log_comm_end():
+                profile_base_info = trim_last_curly_brace(get_profile_info("communication", "E"))
+                end_timestamp = tf.timestamp() * 1e6
+                duration = end_timestamp - self.start_timestamp
+                return tf.print(profile_base_info, ", \"ts\": ", end_timestamp, ", \"duration\": ", duration, "}", sep="", output_stream="file:///profile/pedl_profiling.log")
+
+            def clear_profile_counter():
+                return self.profile_counter.assign(0)
+
+            def increment_profile_counter():
+                return self.profile_counter.assign_add(1)
+
             self._get_gradients_used = True
             self.grads = super(
                 self.__class__, self).get_gradients(loss, params)
             init_aggregation_vars()
             if hvd.size() > 1:
-                if self._aggregation_frequency > 1:
-                    with tf.variable_scope("aggregation_variables", reuse=True):
-                        clear_op = tf.cond(
-                            tf.equal(self.counter, 0), clear_grads, tf.no_op)
-                        with tf.control_dependencies([clear_op]):
-                            aggregation_ops_list = aggregate_grads()
-
-                    aggregation_ops = tf.group(*aggregation_ops_list)
-                    with tf.control_dependencies([aggregation_ops]):
-                        update_ops = [self.counter.assign_add(tf.constant(1))]
-                else:
-                    update_ops = []
+                with tf.variable_scope("aggregation_variables", reuse=True):
```

Comment:
Do we need this variable scope here? And if we do, we should probably not have nested calls to it (L241).

Reply:
Not positive if we do, but technically there are no nested calls to it, the scope is only used for line 238